feat(post): why php in 2019 - ¿por qué php en 2019? (#8)
LinioIT/linioit.github.io: _posts/2019-02-25-why-php-in-2019.md

> ---

Reviewer: It's just an opinion, but... I think when you are writing an article or blog, should not be very "informal" ... using for example `wasn't`, `it's`, etc. 

oh, and there is a lot of unnecessary extra spaces between words.

BTW, the article have a very interesting perspective.

layout: post
title: "Why PHP in 2019?"
date: 2019-02-25 09:40:00 -0500
categories: programming
author: "Roberto Briones"
lang: en
---

> There are a lot of programming languages in the IT environment, but just a few with so much bad reputation like PHP, but is his reputation accurate? And why [one of the most used programming languages](https://haydenjames.io/80-percent-web-powered-by-php/) in the world is at the same time so hated?  🤔 Let’s find out. 🕵🏻‍♂️

Reviewer 2: ```suggestion
There are a lot of programming languages in the IT environment, but just a few with such a bad reputation like PHP, but is his reputation accurate? And why [one of the most used programming languages](https://haydenjames.io/80-percent-web-powered-by-php/) in the world is at the same time so hated?  🤔 Let’s find out. 🕵🏻‍♂️
```

Reviewer 3: Also:

```suggestion
There are a lot of programming languages in the IT environment, but just a few with such a bad reputation like PHP, but is its reputation the accurate one? And why [one of the most used programming languages](https://haydenjames.io/80-percent-web-powered-by-php/) in the world is at the same time so hated?  🤔 Let’s find out. 🕵🏻‍♂️
```

> There are a lot of programming languages in the IT environment, but just a few with so much bad reputation like PHP, but is his reputation accurate? And why [one of the most used programming languages](https://haydenjames.io/80-percent-web-powered-by-php/) in the world is at the same time so hated?  🤔 Let’s find out. 🕵🏻‍♂️

Reviewer 4: The above suggestions are good, but I think they miss the crucial fault with the opening paragraph.

The first sentence is a bit of a run-on. Two times the word "but" is used after a comma. I would try to break it up into something like this:

```suggestion
There are a lot of programming languages in the IT environment, yet few of them have as bad a reputation as PHP. Is this accurate? And if so, why [is one of the _most used_ programming languages](https://haydenjames.io/80-percent-web-powered-by-php/) in the world so hated?  🤔 Let’s find out. 🕵🏻‍♂️
```

I would even consider removing the second sentence:

> Is this accurate?

And continue the paragraph like this:

```suggestion
There are a lot of programming languages in the IT environment, yet few of them have as bad a reputation as PHP. Why [is one of the _most used_ programming languages](https://haydenjames.io/80-percent-web-powered-by-php/) in the world so hated?  🤔 Let’s find out. 🕵🏻‍♂️
```

However this removes the question as to whether or not the first sentence is a true statement. It changes the entire voice/perspective of the post by assuming that, yes, PHP does have a bad reputation.

I'll leave this to the discretion of the author. 😄

## A little of history

> PHP borns in 1995 by the hand of the programmer [Rasmus Lerdorf](https://en.wikipedia.org/wiki/Rasmus_Lerdorf)  , his goal wasn't too ambicious, at the moment he just wanted to have a set of scripts for tracking the visits of his own site; his online resume, that’s kind of explain the reason of it’s first name: ´Personal Home Page Tools´ (PHP Tools), yep, you are right, PHP was not intended to be a programming language at all, but just a set of tools, in fact in the [PHP official announcement](https://groups.google.com/forum/#!msg/comp.infosystems.www.authoring.cgi/PyJ25gZ6z7A/M9FkTUVDfcwJ) this was presented as: *“a set of small tight cgi binaries written in C.”* 

Reviewer 2: ```suggestion
PHP borns in 1995 by the hand of the programmer [Rasmus Lerdorf](https://en.wikipedia.org/wiki/Rasmus_Lerdorf), his goal wasn't too ambitious, at the moment he just wanted to have a set of scripts for tracking the visits of his own site; his online resume, that’s kind of explain the reason of it’s first name: ´Personal Home Page Tools´ (PHP Tools), yep, you are right, PHP was not intended to be a programming language at all, but just a set of tools, in fact in the [PHP official announcement](https://groups.google.com/forum/#!msg/comp.infosystems.www.authoring.cgi/PyJ25gZ6z7A/M9FkTUVDfcwJ) this was presented as: *“a set of small tight cgi binaries written in C.”* 
```

> PHP borns in 1995 by the hand of the programmer [Rasmus Lerdorf](https://en.wikipedia.org/wiki/Rasmus_Lerdorf)  , his goal wasn't too ambicious, at the moment he just wanted to have a set of scripts for tracking the visits of his own site; his online resume, that’s kind of explain the reason of it’s first name: ´Personal Home Page Tools´ (PHP Tools), yep, you are right, PHP was not intended to be a programming language at all, but just a set of tools, in fact in the [PHP official announcement](https://groups.google.com/forum/#!msg/comp.infosystems.www.authoring.cgi/PyJ25gZ6z7A/M9FkTUVDfcwJ) this was presented as: *“a set of small tight cgi binaries written in C.”* 

Reviewer 4: ```suggestion
PHP was born in 1995 by the hand of the programmer [Rasmus Lerdorf](https://en.wikipedia.org/wiki/Rasmus_Lerdorf). His goal wasn't too ambitious. At the moment he just wanted to have a set of scripts for tracking the visits of his online resume. This explains PHP's first name: ´Personal Home Page Tools´ (PHP Tools). PHP was not intended to be a programming language at all, but just a set of tools. In fact, in the [PHP official announcement](https://groups.google.com/forum/#!msg/comp.infosystems.www.authoring.cgi/PyJ25gZ6z7A/M9FkTUVDfcwJ) it was presented as: *“a set of small tight cgi binaries written in C.”* 
```

> But despite the fact that it was not intended to be a programming language it quickly got some adopters, quite some actually, one of the main reasons for that is because at the time there wasn’t a lot of programming languages that could be used on Web Development (remember that the first web browser Mosaic just born in 1993, so the Internet was very new) and those that already existed like Ruby or Python required a bigger learning curve to create a simple webpage, also this languages were general purpose languages that means that could be used to develop different types of applications, and PHP was thought as a series of tools for Web Development since the beginning.

Reviewer 4: ```suggestion
Despite the fact that it was not intended to be a programming language, it quickly gained adoption. One of the main reasons was because there wasn’t a lot of programming languages that could be used for Web Development (the first web browser, Mosaic, was just born in 1993, so the Internet was still new). Existing languages like Ruby or Python required a bigger learning curve to create a simple webpage, and they were general purpose languages used to develop different types of applications. PHP was meant for Web Development from the beginning.
```

Another thing that helped PHP to become one of the most used programming languages for Web Development is the boom of the dot-com bubble, when the companies came across the fact that if you were not on the internet, they practically did not exist. 

> With the time and a lot of help of the growing community, PHP "bit by bit" become without any doubt into a general purpose programming language with a simple way to develop Websites and it’s name evolves as we know it today,  a recursive acronym that means: ***PHP: Hypertext Preprocessor*** 

Reviewer 4: ```suggestion
With time and help from the growing community, PHP became a general purpose programming language with a simple way to develop Websites. Its name evolved as we know it today, a recursive acronym that means: ***PHP: Hypertext Preprocessor*** 
```

> ## Okey, but if it’s so great, how come it’s so hated?

Reviewer 4: Okey should be spelled => Okay.

> ## Okey, but if it’s so great, how come it’s so hated?

Author: This tries to reference the Simpsons phrase: "If he's so smart, how come he's dead?" 
![g225bwes4g521](https://user-images.githubusercontent.com/1811438/54693330-08afff80-4aec-11e9-8abb-fa7b6d158f8a.jpg)

Maybe adding the image? 😅

> ## Okey, but if it’s so great, how come it’s so hated?

Reviewer 5: I would try to keep the content less casual.

> As I mentioned before, PHP wasn’t mean to be a Programming Language, that’s very important to know, because that means that in order to become one it needed to be adapted by the creator along with the community, and like all developers know; **planning afterwards always means trouble.**

Reviewer 4: ```suggestion
As I mentioned before, PHP wasn’t mean to be a Programming Language. In order to become one it needed to be adapted by the creator along with the community. Like all developers know, **planning afterwards always means trouble.**
```

> This troubles was translated into deficiencies if we compare it with other programming languages that were planned to work well from the beginning.

Reviewer 4: ```suggestion
These troubles were translated into deficiencies compared to other programming languages that were planned to work well from the beginning.
```

> One of the most common critics for PHP is it inconsistency, but not take my word for it, let me show you an example:

Reviewer 4: ```suggestion
One of PHP's most common criticisms is its inconsistency, but don't take take my word for it. Let me show you an example:
```

---
layout: post
title: "Why PHP in 2019?"
date: 2019-02-25 09:40:00 -0500
categories: programming
author: "Roberto Briones"
lang: en
---

> There are a lot of programming languages in the IT environment, yet few of them have as adverse a reputation as PHP. Is this accurate? And if so, why [is one of the _most used_ programming languages](https://haydenjames.io/80-percent-web-powered-by-php/) in the world so hated? 🤔 Let’s find out. 🕵🏻‍♂️

Reviewer 6: Prefer "criticized" instead of "hated"

## A little of history

PHP was born in 1995 by the hand of the programmer [Rasmus Lerdorf](https://en.wikipedia.org/wiki/Rasmus_Lerdorf). His goal wasn't too ambitious. He wanted to have a set of scripts for tracking the visits of his online resume. This explains PHP's first name: ´Personal Home Page Tools´ (PHP Tools). PHP was not intended to be a programming language at all, but a set of tools. In fact, in the [PHP official announcement](https://groups.google.com/forum/#!msg/comp.infosystems.www.authoring.cgi/PyJ25gZ6z7A/M9FkTUVDfcwJ) it was presented as *“a set of small tight CGI binaries written in C.”*  

> Even though it was not intended to be a programming language, it quickly gained adoption. One of the main reasons was that there wasn’t a lot of programming languages that could be used for Web Development (the first web browser, Mosaic, was just born in 1993, so the Internet was still new). Existing languages like Ruby or Python required a bigger learning curve to create a simple webpage. They were also general-purpose languages used to develop different types of applications. PHP was meant for Web Development from the beginning.

Reviewer 6: Prefer "steeper" instead of "bigger"

Another thing that helped PHP to become one of the most used programming languages for Web Development was the boom of the dot-com bubble. Companies began to realize that they required an internet presence. 

With time and help from the growing community, PHP became a general-purpose programming language with a simple way to develop Websites. Its name evolved as we know it today, a recursive acronym that means ***PHP: Hypertext Preprocessor***.

## Okay, but if it’s so great, how did it become so controversial?

> As I mentioned before, PHP was not meant to be a Programming Language. . In order to become one, it had to be modified by its creator, along with its community. Like all developers know, **planning afterward always means trouble.**

Reviewer 7: ```suggestion
As I mentioned before, PHP was not meant to be a Programming Language. In order to become one, it had to be modified by its creator, along with its community. Like all developers know, **planning afterward always means trouble.**
```

Author: Thanks!

> As I mentioned before, PHP was not meant to be a Programming Language. This is something very important to keep in mind, because that means that, to become one, it needed to be adapted by the creator along with the community. Like all developers know, **planning afterward always means trouble.**

Reviewer 6: I would review this sentence again.

As I mentioned before, PHP was not meant to be a Programming Language. In order to become one, it had to be modified by its creator, along with its community. Like all developers know, **planning afterward always means trouble.**

These troubles were translated into deficiencies compared to other programming languages that were planned to work well from the beginning.

One of PHP's most common criticisms is its inconsistency, but don't take my word for it. Let me show you an example:

If we explore the naming of the native functions that [the language give us for string management](http://php.net/manual/es/ref.strings.php) we will find functions like these:

`str_replace`
`strpos `
`substr `

> All three of them are mean to manipulate strings. The first function has ***“str”*** as a prefix and a ***"_"*** to separate the words. The second uses the ***“str”*** prefix but doesn’t use the underscore to separate the words. And the last one doesn’t even use the ***“str”*** prefix or underscore at all. Instead it uses a postfix. 

Reviewer 8: All three of them are **meant** to manipulate strings...

> Let's take a look, all three of them are mean to manipulate strings. The first function has ***“str”*** as prefix and a ***"_"*** underscore to separate the words. The second one it still uses the ***“str”*** prefix but doesn’t use the underscore to separate the words. And the last one doesn’t even use the ***“str”*** prefix but it is used as postfix and it doesn’t use the underscore either. 

Reviewer 4: ```suggestion
All three of them are mean to manipulate strings. The first function has ***“str”*** as a prefix and a ***"_"*** to separate the words. The second uses the ***“str”*** prefix but doesn’t use the underscore to separate the words. And the last one doesn’t even use the ***“str”*** prefix or underscore at all. Instead it uses a postfix.
```

> Now imagine that you need to guess the name of a function that searches a string into another string. In which convention that would be? Now you get the point? 

Reviewer 4: ```suggestion
Now imagine that you need to guess the name of a function that searches a string into another string. In which convention would that be? Now do you get the point? 
```

> As you can see the inconsistency is clear. That kind of stuff happens with a lot of the internal functions of the language, not just with the ones that manipulate strings. That is kind of challenging when you try to remember which function to use. It also makes it difficult to be productive, requiring you to check the documentation over and over.

Reviewer 6: Prefer "pattern" instead of "stuff".

Use "over and over again"

> As you can see the inconsistency is clear. That kind of stuff happens with a lot of the internal functions of the language, not just with the ones that manipulate strings and that is kind of challenging when you try to remember which function use. It also makes difficult to get into development without checking the documentation a lot of times.

Reviewer 4: ```suggestion
As you can see the inconsistency is clear. That kind of stuff happens with a lot of the internal functions of the language, not just with the ones that manipulate strings. That is kind of challenging when you try to remember which function to use. It also makes it difficult to be productive, requiring you to check the documentation over and over.
```

> Another downside was the leak of *"basic stuff"* that other programming languages provide from the beginning, like typed properties [(hopefully we will have it in PHP 7.4)](https://wiki.php.net/rfc/typed_properties_v2). Or like an Object Oriented Programming paradigm out the box as Java does (you can't develop in Java if it's not Object Oriented). The return type hinting [(recently added in PHP 7.1)](https://secure.php.net/manual/en/migration71.new-features.php).

Reviewer 6: Prefer "such as" instead of "like"

> Another downside was the leak of *"basic stuff"* that other programming languages provide from the beginning, like typed properties [(hopefully we will have it in PHP 7.4)](https://wiki.php.net/rfc/typed_properties_v2). Or like an Oriented Object Programming paradigm out the box as Java does (you can't develop in Java if it's not Object Oriented). The return type hinting [(recently added in PHP 7.1)](https://secure.php.net/manual/en/migration71.new-features.php).

Reviewer 4: > Oriented Object Programming

Should this be Object Oriented Programming?

Or the lack of native data structures in older versions, the only native data structure in PHP was the array [that is, in reality, a Hash table](https://nikic.github.io/2012/03/28/Understanding-PHPs-internal-array-implementation.html). But that means that you needed to create your own data structures from scratch when you needed it. Now thanks to [SPL classes](http://php.net/manual/es/spl.datastructures.php) we have more options.

> All those kind of disadvantages are the main reasons (but not the only ones) that explain why PHP is so hated by the purist of developers and fans of other languages.

Reviewer 6: "kinds" instead of "kind"

Prefer "disliked" instead of "hated"

> All those kind of disadvantages are the main reasons (but not the only ones) that explain why PHP is so hated by the most purist developers and fans of other languages. 

Reviewer 4: ```suggestion
All those kind of disadvantages are the main reasons (but not the only ones) that explain why PHP is so hated by the purist of developers and fans of other languages. 
```

Also if we take in count that since PHP was born until now a lot of new programming languages were created with more awareness about the details and the actual needs of Web Development, you can understand why all the hate towards it.

> ## Then, why using it in 2019?

Reviewer 4: ```suggestion
## Then why use it in 2019?
```

Well, a lot of things happened since those days. Some of the concerns that the language had are still there, like the naming inconsistency. But a lot of others are not an issue anymore. 

> In the beginning, everyone used to programming in PHP as they wanted to do it. Since PHP doesn’t impose you a paradigm to do it, you could write an entire website in a single file, or you could do it with OOP, creating many files and Classes as you need. Also if you needed some third party library you needed to download it manually and add it to your project.

Reviewer 4: ```suggestion
In the beginning, everyone used to program in PHP they way they wanted to. Since PHP doesn’t impose you a paradigm to do it, you could write an entire website in a single file, or you could do it with OOP, creating many files and classes as you need. Also, if you needed some third party library, you needed to download it manually and add it to your project.
```

> Since then [PHP-FIG](https://www.php-fig.org/) born! FIG stands for _“Framework Interoperability Group”_. They are a group of people trying to figure out some standards for Modern PHP Development. They do that through PSR’s (PHP Standard Recommendations) that helps the developers to use some standards in their applications and then turn it into an efficient and modern web application that will be very similar to others using the same standards.

Reviewer 4: ```suggestion
Since then, [PHP-FIG](https://www.php-fig.org/) was born! FIG stands for _“Framework Interoperability Group”_. They are a group of people trying to figure out some standards for modern PHP development. They do that through PSR’s (PHP Standard Recommendations) that help developers use some standards in their applications and then turn it into an efficient and modern web application that will be very similar to others using the same standards.
```

> Also now we have Frameworks (a lot of them), like ***Symfony, Laravel, CakePHP, Zend, Yii,*** etc. Each one trying to give us a way to develop modern Web Applications based in PHP. In my opinion, these frameworks *were the key* for giving PHP a second breath. Thanks to them now we can develop in a faster, cleaner, and powerful way from tiny to big applications. 

Reviewer 4: ```suggestion
Also, now we have Frameworks (a lot of them), like ***Symfony, Laravel, CakePHP, Zend, Yii,*** etc. Each one tries to give us a way to develop modern web applications based in PHP. In my opinion, these frameworks *were the key* for giving PHP a second breath. Thanks to them now we can develop in a faster, cleaner, and powerful way from tiny to big applications. 
```

> We also have now really useful tools like Composer, that resolves the problem of managing dependencies in a smart way, allowing us to have the latest version of a certain third-party library or even lock it to a specific version. It also helps us to have a big repository of third-party libraries that any developer can use easily.

Reviewer 4: ```suggestion
We also now have really useful tools like Composer, that resolves the problem of managing dependencies in a smart way, allowing us to have the latest version of a certain third-party library or even lock it to a specific version. It also helps us to have a big repository of third-party libraries that any developer can use easily.
```

> According to [TIOBE Index](https://www.tiobe.com/tiobe-index/) (an index of the *most popular Programming Languages in the World*). At the moment of writing this (February 2019) PHP is the 9th Language in the list, just behind C# and just above SQL and leaving behind languages like Swift (20th), Go (18th), or Ruby (16th). So clearly PHP is not dead at all, on the contrary, I think that it’s future it’s very promising. 

Reviewer 4: ```suggestion
According to [TIOBE Index](https://www.tiobe.com/tiobe-index/), an index of the *most popular Programming Languages in the World*, at the moment of writing this (February 2019) PHP is the 9th Language in the list, just behind C# and just above SQL and leaving behind languages like Swift (20th), Go (18th), and Ruby (16th). So clearly PHP is not dead at all. On the contrary, I think that it’s future is very promising. 
```

Also, you can check the [list of the most popular websites in the world](https://en.wikipedia.org/wiki/Programming_languages_used_in_most_popular_websites) and notice that at least 5 of them uses PHP in one way or another.

PHP is still one of the recommended programming languages to learn this year by [several sites](https://hackr.io/blog/best-programming-languages-to-learn-2019-jobs-future) and [PHP is used by 78.9% of all the websites whose server-side programming language we know](https://w3techs.com/technologies/details/pl-php/all/all).

## Conclusion

> PHP is far from being the best programming language. It was designed as a set of scripts that evolved into a programming language and now has evolved to become one of the most used for web development. And like any evolution, it is constantly adapting to the changes and to the needs of developers. It tries to become better and better, sometimes even taking advantage of some of the features that are already implemented by other programming languages, but that is just part of the evolution.

Reviewer 6: We can't say this sentence: "PHP is far from being the best programming language", otherwise why is Linio not using the best programming language? :)

Reviewer 9: Linio does not use the best programming language because doesn't exist, I think @Briones put that just as an expression, programming languages are using for different objectives it depends on what do you want, and that's what he describes in the next lines, in my opinion is not a big deal have that expression in our blog but I don't know what is the #engineers opinion

cc @LinioIT/backend-services

Author: @anthonybensadoun I think the same, there is not a best programming language, is a programming language is a tool, I also think we should be honest and less politically correct, PHP is not just "disliked", is indeed hated, we have posts like this in internet and is also well known in the community. 

https://whydoesitsuck.com/why-does-php-suck/

https://blog.codinghorror.com/php-sucks-but-it-doesnt-matter/

https://eev.ee/blog/2012/04/09/php-a-fractal-of-bad-design/

The point for me is trying to emphatize but that ideollogy and then show them that there is no reason for hate, and can be used, and should be used depending of the needs. Don't know if trying to mask the reality should be the way to go.

Reviewer 5: The problem is not the idea, it's the wording. When you structure the statement like:

> far from being the best programming language

You are implying that such a thing exists. This is enforced by the usage of definite article **the** followed by a superlative, **best**. Based on your comment, your statement clearly conflicts with your belief that a best programming language doesn't exist.

You're probably better phrasing it like:

> although PHP is far from an idealistic enterprise programming language

or 

> PHP is far from the traditional programming language design standards

Either way, you can use that to continue emphasizing the pros and cons of the organic design nature of the language, and maybe even present an argument comparing the advantages and disadvantages of design by committee in free and open source communities.

Reviewer 6: Right, the statement implies that a best programming language exists. I think your suggestions Klaus work much better.

@Briones the point is not to mask anything, if we are convinced that PHP is the right choice for Linio, then we simply can't say that it is far from being the best option, those statements conflict. If we were a car manufacturer, and wanted to build the best car possible, we would not say that the steel we use to build the car is far from the best option. The conclusion would say that all steels have their pros and cons, but we have picked this one for a good reason, and it is the right choice to build the car we are building. Do you follow?

Author: Gotcha, thanks both, can you review the latest changes that I did in the wording?

> PHP is far for being the best programming language. It was designed as a set of scripts that evolved into a programming language and now has evolved to become one of the most used for web development. And like in any evolution it is constantly adapting to the changes, to the developer needs. It tries to become better and better, sometimes even taking advantage of some of the features that are already implemented by other programming languages, but that is just part of the evolution.

Reviewer 4: ```suggestion
PHP is far from being the best programming language. It was designed as a set of scripts that evolved into a programming language and now has evolved to become one of the most used for web development. And like any evolution, it is constantly adapting to the changes and to the needs of developers. It tries to become better and better, sometimes even taking advantage of some of the features that are already implemented by other programming languages, but that is just part of the evolution.
```

> The main problem with PHP is not PHP at all, is us, *the developers*. That from the beginning were trying to write applications without any concern about the standards, or the maintainability, without focusing in unit testing, or even in readability of the code that we made.

Reviewer 4: ```suggestion
The main problem with PHP is not PHP at all, but us, *the developers*. From the beginning we're trying to write applications without any concern about the standards, or the maintainability, without focusing in unit testing, or even in readability of the code that we made.
```

> All of that resulted in a lot of old (and some new) Web Applications that right now seems to be unmaintainable, without structure, with PHP code mixed with HTML tags and Javascript or CSS code… This is the reason for the deepened the unfounded hate of PHP. 

Reviewer 8: All of that resulted in a lot of old (and some new) Web Applications that right now **seem** to be unmaintainable

Reviewer 8: This is the reason for the deepened ~the~ unfounded hate of PHP.

> Right now PHP is a lot of times the ideal programming language for some **Modern Web Applications**. With it, we can generate code that will be maintainable, robust, fast, clean, and will easily connect with other popular technologies. 

Reviewer 8: Right now, PHP is...

> Here at **Linio** we use PHP since the beginning for numerous projects. And right now we are still using it for upcoming projects and we are convinced that PHP is a good choice, as good as any other.

Reviewer 4: ```suggestion
Here at **Linio** we use PHP for numerous projects. And right now we are still using it for upcoming projects and we are convinced that PHP is a good choice, as good as any other.
```
